File: android/app/src/main/java/app/herawallet/client/Haptica.java
package app.herawallet.client;

import android.content.Context;
import android.os.Build;
import android.os.CombinedVibration;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.os.VibratorManager;
import android.webkit.JavascriptInterface;

/**
 * Vibración de respuesta al tacto.
 *
 * Los pulsos son cortos y flojos a propósito. Una vibración larga se percibe
 * como un aviso —algo va mal, algo llega—, y aquí solo se está confirmando que
 * el dedo tocó donde quería. Las apps que se sienten bien vibran poco y breve.
 *
 * La web llama a esto por `HeraVibrar`: el `navigator.vibrate` del WebView no
 * es fiable y no distingue intensidades.
 */
public class Haptica {

    /** Confirmación de un toque. Lo más suave que se puede sentir. */
    public static final int TOQUE = 12;
    /** Algo empieza o termina: grabar, abrir una hoja. */
    public static final int PULSO = 22;
    /** Algo salió bien y merece notarse. */
    public static final int EXITO = 34;

    private final Context contexto;

    public Haptica(Context contexto) {
        this.contexto = contexto;
    }

    private static Vibrator vibrador(Context contexto) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            VibratorManager gestor =
                    (VibratorManager) contexto.getSystemService(Context.VIBRATOR_MANAGER_SERVICE);
            return gestor == null ? null : gestor.getDefaultVibrator();
        }
        return (Vibrator) contexto.getSystemService(Context.VIBRATOR_SERVICE);
    }

    private static void vibrar(Context contexto, int milisegundos, int amplitud) {
        try {
            Vibrator v = vibrador(contexto);
            // Un teléfono sin motor no debe llegar aquí con un error.
            if (v == null || !v.hasVibrator()) return;

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                v.vibrate(VibrationEffect.createOneShot(milisegundos, amplitud));
            } else {
                v.vibrate(milisegundos);
            }
        } catch (Exception ignorado) {
            // La vibración es un adorno: nunca puede tumbar una acción.
        }
    }

    public static void toque(Context contexto) {
        vibrar(contexto, TOQUE, 60);
    }

    public static void pulsar(Context contexto) {
        vibrar(contexto, PULSO, 120);
    }

    public static void exito(Context contexto) {
        vibrar(contexto, EXITO, 170);
    }

    // --- Lo que llama la web -------------------------------------------------

    @JavascriptInterface
    public void toque() {
        toque(contexto);
    }

    @JavascriptInterface
    public void pulso() {
        pulsar(contexto);
    }

    @JavascriptInterface
    public void exito() {
        exito(contexto);
    }
}

File: src/haptica.ts
/**
 * Vibración de respuesta al tacto.
 *
 * Los pulsos son cortos y flojos a propósito. Una vibración larga se lee como
 * un aviso —algo va mal, algo llega—, y aquí solo se confirma que el dedo tocó
 * donde quería. Las apps que se sienten bien vibran poco y breve.
 *
 * Dentro de la app se usa el puente nativo, que distingue intensidades. En el
 * navegador se cae a `navigator.vibrate`, que solo entiende duraciones y no
 * existe en iPhone: allí, sencillamente, no vibra nada.
 */

type Intensidad = 'toque' | 'pulso' | 'exito';

/** Duraciones para el respaldo del navegador, en milisegundos. */
const DURACIONES: Record<Intensidad, number> = {
  toque: 10,
  pulso: 20,
  exito: 30,
};

interface PuenteHaptica {
  toque: () => void;
  pulso: () => void;
  exito: () => void;
}

const puente = (): PuenteHaptica | null => {
  const p = (window as any)?.HeraVibrar;
  return p && typeof p.toque === 'function' ? p : null;
};

/**
 * Quien pidió al sistema que no le muevan la interfaz tampoco quiere que le
 * vibre en la mano.
 */
const quieto = () =>
  typeof window !== 'undefined' &&
  window.matchMedia?.('(prefers-reduced-motion: reduce)').matches;

function vibrar(intensidad: Intensidad) {
  if (typeof window === 'undefined' || quieto()) return;

  try {
    const nativo = puente();
    if (nativo) {
      nativo[intensidad]();
      return;
    }

    navigator.vibrate?.(DURACIONES[intensidad]);
  } catch {
    // Vibrar es un adorno: no puede tumbar la acción que lo disparó.
  }
}

/** Confirmación de un toque. Lo más suave que se puede sentir. */
export const toque = () => vibrar('toque');

/** Algo empieza o termina: se abre una hoja, arranca una grabación. */
export const pulso = () => vibrar('pulso');

/** Algo salió bien y merece notarse: un registro guardado, un logro. */
export const exito = () => vibrar('exito');

/**
 * Vibración en cualquier elemento pulsable de la app, sin tener que tocar cada
 * botón uno por uno.
 *
 * Se escucha en la fase de captura y sobre `pointerdown`, no sobre `click`: la
 * respuesta tiene que llegar con el dedo, no cuando se levanta. Se filtra a
 * botones y enlaces reales para no vibrar al desplazarse o al seleccionar
 * texto.
 */
export function activarHapticaGlobal() {
  if (typeof document === 'undefined') return;

  document.addEventListener('pointerdown', (evento) => {
    const destino = evento.target as HTMLElement | null;
    if (!destino?.closest) return;

    const pulsable = destino.closest(
      'button, a[href], [role="button"], [role="switch"], [role="tab"], input[type="range"]'
    ) as HTMLElement | null;

    if (!pulsable) return;
    if (pulsable.hasAttribute('disabled') || pulsable.getAttribute('aria-disabled') === 'true') return;

    toque();
  }, { capture: true, passive: true });
}
